lib/adm.js

```diff
@@ -2385,36 +2394,306 @@ maAdm.prototype.dumpConfigLab = function(options, callback) {
+                            ctx.setupServerFromName[s.hostname] = s;
+                        }
+
+                        next();
+                    });
+                },
+
+                // Check the server options and normalize them to arrays of server
+                // UUIDs.
+                function ensureServerOpts(ctx, next) {
+                    if (!options.servers || options.servers.length === 0) {
+                        ctx.serverUuids = ctx.setupServers.map(function(s) {
+                            return s.uuid;
+                        });
+                    } else {
+                        ctx.serverUuids = [];
+                        for (let uuidOrHostname of options.servers) {
+                            let s = ctx.setupServerFromName[uuidOrHostname];
+                            if (!s) {
+                                next(
+                                    new VError(
+                                        '"%s" is not the UUID or hostname of a setup server',
+                                        uuidOrHostname
+                                    )
+                                );
+                                return;
+                            }
+                            ctx.serverUuids.push(s.uuid);
+                        }
+                    }
+
+                    if (
+                        !options.storageServers ||
+                        options.storageServers.length === 0
+                    ) {
+                        ctx.storageServerUuids = ctx.serverUuids.slice();
+                    } else {
+                        ctx.storageServerUuids = [];
+                        for (let uuidOrHostname of options.storageServers) {
+                            let s = ctx.setupServerFromName[uuidOrHostname];
+                            if (!s) {
+                                next(
+                                    new VError(
+                                        '"%s" is not the UUID or hostname of a setup server',
+                                        uuidOrHostname
+                                    )
+                                );
+                                return;
+                            }
+                            ctx.storageServerUuids.push(s.uuid);
+                        }
+                    }
+
+                    next();
+                },
+
+                // Which server for instances? Some guidelines:
+                // - Spread "storage" insts across all "storageServers" and warn if
+                //   there aren't enough servers.
+                // - Spread other insts of a given service across all "servers" and
+                //   warn if there aren't enough servers.
+                // - Bias (for now) to having one of everything on the headnode to
+                //   facilitate dev/debugging, if the headnode is one of the
+                //   servers.
+                function createLayout(ctx, next) {
```

Comment:
> I think I would expect that if the number of storage servers is less than CNs + HN, that the HN _does not_ get a storage instance

That would be a nice thing to do, but this change doesn't tackle that at all.


> I guess this is just generating the config though so the operator can always fix that themselves.

Sorta. If used as part of `manta-deploy-lab` it will go ahead and create that storage zone on the headnode, potentially.

Anyway, I'm calling this one out of scope.



> Overall I find this function kinda long and hard to follow and probably could at least use a comment explaining its algorithms.

Thanks, yes I'll try to document it.